File: Debugging/Arithmetic/addition.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
    long n = 1000000, i;
    double result1 = 1.0, result2 = 0.0, number;
    if (argc > 1)
        n = atol(argv[1]);
    for (i = 0; i < n; i++) {
        number = (1.0e-17*rand())/RAND_MAX;
        result1 += number;
        result2 += number;
    }
    result2 += 1.0;
    printf("1 + sum(x) = %.15le\n", result1);
    printf("sum(x) + 1 = %.15le\n", result2);
    printf("relative error = %.3le\n", (result2 - result1)/result1);
    return EXIT_SUCCESS;
}

File: Debugging/Arithmetic/addition.py
#!/usr/bin/env python

from argparse import ArgumentParser
import numpy as np

if __name__ == '__main__':
    arg_parser = ArgumentParser(description='illustrate addition accuracy')
    arg_parser.add_argument('--numbers', type=int, default=1000000,
                            help='array size')
    options = arg_parser.parse_args()
    array = 1.0e-16*np.random.uniform(size=options.numbers)
    result1 = 1.0
    result2 = 0.0
    for x in array:
        result1 += x
        result2 += x
    result2 += 1.0
    print(f'1 + sum(x) = {result1}')
    print(f'sum(x) + 1 = {result2}')
    dev = (result2 - result1)/result1
    print(f'relative error = {dev:.3e}')
